Cleanup & document pipeline manager API.

ryzhyk opened the pull request:
Use the `utoipa` crate to generate OpenAPI spec for the pipeline manager API, and Swagger documentation for it, available on `/api-doc/openapi.json` and `/swagger-ui/` endpoints respectively.

Unlike OpenAPI generators, `utoipa` doesn't automatically generate server-side stubs for the API, but rather documents an existing, manually implemented, API.  Although it does interact with the Rust type system, it is not 100% type-safe, i.e., the generated API may not match the actual input and output formats used by the server.  Unfortunately, this appears to be the best option the Rust ecosystem offers today.

While at it, I also cleaned up the API in a couple of ways:

- Use JSON for all inputs and outputs (as opposed to returning plain text strings in some cases).  In particular, errors are now returned as JSON:

  ``` { "message": "Cannot delete a project while some of its pipelines are running" } ```

- Convert `list_project_configs` and `list_project_pipelines` into `GET` methods that take project id as URL-encoded parameter for consistency with `project_status`, `project_code`, etc.

@mbudiu-vmw, the preferred way to work with this API from the browser would be to generate client-side typescript bindings for it, but it's probably not worth bothering with now, since we will eventually re-implement the UI anyway.

lalithsuresh commented on pipeline_manager/src/main.rs:
```diff
@@ -367,25 +482,36 @@ async fn project_status(state: WebData<ServerState>, req: HttpRequest) -> impl R
+    ),
+    tag = "Project"
+)]
 #[post("/new_project")]
```
`POST /projects`

lalithsuresh commented on pipeline_manager/src/main.rs:
```diff
@@ -475,12 +608,21 @@ struct CompileProjectRequest {
+    ),
+    tag = "Project"
+)]
 #[post("/compile_project")]
```
`POST /projects/{project_id}/compile`

ryzhyk replied:
Isn't it better to do `POST /projects/compile` and extract `project_id` from JSON?  Otherwise, we have a mix of url-encoded and JSON arguments.

lalithsuresh replied:
Yes, I made a similar suggestion to that effect in another comment now that I understand the workflow better.

lalithsuresh commented on pipeline_manager/src/main.rs:
```diff
@@ -509,13 +651,21 @@ struct CancelProjectRequest {
+    tag = "Project"
+)]
 #[post("/cancel_project")]
 async fn cancel_project(
```
What's the difference between canceling a project and deleting it?

ryzhyk replied:
`/cancel` cancels compilation project compilation.  The project remains in the database and can be compiled/modified/queried.

lalithsuresh replied:
To follow a create/<modify>/delete cycle for a project, here is an approach that comes to mind.

* `POST /projects/`     # creates a project, not yet compiled
* `GET /projects/{project_id}`
* `POST /compiler/{project_id}` # creates a compilation job
* `GET /compiler/{project_id}` # gets the status of the compilation job
* `DELETE /compiler/{project_id}` # ends the compilation job

An alternative that some people use, which might be overkill for now:

* `POST /projects/`     # creates a project, not yet compiled
* `GET /projects/{project_id}`
* `PUT/PATCH /projects/{project_id}`   # modify a "compilation" field in the project that changes its desired state, which triggers compilation. AWS APIs have this approach to attaching "Action" sub-resources to resources.

lalithsuresh commented on pipeline_manager/src/main.rs:
```diff
@@ -588,10 +747,17 @@ struct NewConfigResponse {
+    ),
+    tag = "Config"
+)]
 #[post("/new_config")]
```
`POST /configs/`

ryzhyk replied:
So a create operation usually does not use a verb, i.e., not `/configs/new`?

lalithsuresh replied:
Yup. POST is the verb, which usually creates something.

lalithsuresh commented on pipeline_manager/src/main.rs:
```diff
@@ -636,10 +802,17 @@ struct UpdateConfigResponse {
+    ),
+    tag = "Config"
+)]
 #[post("/update_config")]
```
`POST /configs/{config_id}` or maybe even PATCH

lalithsuresh commented on pipeline_manager/src/main.rs:
```diff
@@ -661,17 +834,25 @@ async fn update_config(
+    ),
+    tag = "Config"
+)]
 #[post("/delete_config")]
```
`DELETE /configs/{config_id}`

lalithsuresh commented on pipeline_manager/src/main.rs:
```diff
@@ -687,30 +868,34 @@ async fn delete_config(
+    ),
+    tag = "Config"
+)]
+#[get("/list_project_configs/{project_id}")]
```
`GET /projects/{project_id}/configs`

lalithsuresh commented on pipeline_manager/src/main.rs:
```diff
@@ -746,21 +931,32 @@ struct NewPipelineResponse {
+    ),
+    tag = "Pipeline"
+)]
 #[post("/new_pipeline")]
```
`POST /pipeline/`

lalithsuresh commented on pipeline_manager/src/main.rs:
```diff
@@ -773,30 +969,34 @@ async fn new_pipeline(
+    ),
+    tag = "Pipeline"
+)]
+#[get("/list_project_pipelines/{project_id}")]
```
`GET /projects/{project_id}/pipelines`

lalithsuresh commented on pipeline_manager/src/main.rs:
```diff
@@ -823,13 +1023,25 @@ pub(self) struct KillPipelineRequest {
+    ),
+    tag = "Pipeline"
+)]
 #[post("/kill_pipeline")]
```
`DELETE /pipeline/{pipeline_id}` (or POST?) (what's the difference between kill pipeline and delete? is it just stop vs delete?)

ryzhyk replied:
delete = kill + delete from DB.  A killed pipeline can still be inspected until it is deleted.

lalithsuresh replied:
For pipelines, it makes sense to have a status field for the pipeline that we can change with a `PUT` or a `PATCH` to start or stop it. So we'll have:

```
CREATE /pipelines
PUT/PATCH /pipelines/{pipeline_id}
DELETE /pipelines/{pipeline_id}
```

lalithsuresh reviewed (commented):
I've made suggestions for how to keep the API design familiar to folks used to REST. In general, describe resources (nouns) and operations on them (GET, POST, etc.), rather than describing verbs. It's also easier to evolve later.

Could we add more documentation about the accepted configs? For example, I couldn't easily find examples for the file transport.

ryzhyk commented on pipeline_manager/src/runner.rs:
```diff
@@ -441,17 +443,23 @@ scrape_configs:
             db.set_pipeline_killed(pipeline_id).await?;
-            Ok(HttpResponse::Ok().finish())
+            Ok(HttpResponse::Ok()
+                .body(serde_json::to_string("Pipeline successfully terminated.").unwrap()))
```
Same thing.  The API is called `kill` but then I decided that `terminate` is a better user-facing name.  Yeah, `shutdown` is probably better.

ryzhyk commented on pipeline_manager/src/main.rs:
```diff
@@ -327,24 +436,30 @@ struct ProjectStatusResponse {
+    ),
+    tag = "Project"
+)]
 #[get("/project_status/{project_id}")]
```
and `/list_projects` should be just `GET /project`?

lalithsuresh replied:
Yup. `GET /projects` for example.

mbudiu-vmw commented on pipeline_manager/src/db.rs:
```diff
@@ -523,16 +563,16 @@ impl ProjectDB {
             .execute(
                 "INSERT INTO project_config (id, project_id, version, name, config) VALUES($1, $2, 1, $3, $4)",
-                &[&id, &project_id, &config_name, &config],
+                &[&id.0, &project_id.0, &config_name, &config],
```
would a getter be better than accessing .0?

mbudiu-vmw commented on pipeline_manager/src/main.rs:
```diff
@@ -434,13 +560,20 @@ struct UpdateProjectResponse {
-/// # HTTP errors
-///
-/// * `NOT_FOUND` - `project_id` does not exist in the database.
+#[utoipa::path(
```
these don't seem to save you much typing.
Are they mostly for documentation using swagger?

ryzhyk replied:
utoipa generates an OpenAPI spec, which is a piece of JSON that other tools can use to generate client-side bindings, e.g., I'm working on Python bindings now.  And it also generates swagger documentation from that.

lalithsuresh commented on pipeline_manager/scripts/delete_project.sh:
```diff
@@ -9,4 +9,4 @@ if [ "$#" -ne 1 ]; then
 fi
 
-curl -X POST http://localhost:8080/delete_project  -H 'Content-Type: application/json' -d '{"project_id":'$1'}'
+curl -X DELETE http://localhost:8080/projects -H 'Content-Type: application/json' -d '{"project_id":'$1'}'
```
For consistency, it's better to do a delete to `/projects/{project_id}`. Similar issues with delete for pipelines and configs. Given that we already have URLs like `/projects/$1/{config, pipelines}` etc elsewhere.

lalithsuresh replied:
Otherwise, what we're suggesting with the API is a way to delete a collection (e.g. multiple projects), in which case the payload should be able to state an array of IDs to delete. This is also fine, just want to be sure what the API is.

ryzhyk commented:
@lalithsuresh , I fixed the delete apis. I also implemented request forwarding via the main server.  When you click on `Pipelines` to see the list of running pipelines, there should be buttons to retrieve pipeline status, metadata, start and pause the pipeline. Let me know if it works for you, so we can cleanup and merge this.

lalithsuresh commented:
LGTM. I was able to run it end to end with a file based transport example. I'll send a PR for that after we merge this.
